### PR TITLE
Don't unwrap in Kademlia protocol

### PR DESCRIPTION
**Pull request by tomaka**

### CODE REVIEW

File: kad/src/protocol.rs [hunk 1]
```diff
@@ -262,13 +272,22 @@ fn proto_to_msg(mut message: protobuf_structs::dht::Message) -> Result<KadMsg, I
                 Ok(KadMsg::FindNodeReq {
                     key: message.take_key(),
                 })
+
             } else {
+                let mut closer_peers = Vec::with_capacity(message.get_closerPeers().len());
```
Comment (Vurich) [inline kad/src/protocol.rs]: Why not:

```rust
message.mut_closerPeers()
    .iter_mut()
    .filter_map(|peer| Peer::from_peer(peer).ok())
    .collect::<Vec<_>>()
```

Is the possible increased number of allocations really worth it?

File: kad/src/protocol.rs [hunk 2]
```diff
@@ -262,13 +272,22 @@ fn proto_to_msg(mut message: protobuf_structs::dht::Message) -> Result<KadMsg, I
...
+                for peer in message.mut_closerPeers().iter_mut() {
+                    // TODO: for now we don't parse properly the peer, so it is possible that we
+                    //       get parsing errors even though the peer is valid ; we ignore these,
+                    //       but ultimately we should just error altogether
+                    if let Ok(peer) = Peer::from_peer(peer) {
+                        closer_peers.push(peer);
+                    }
+                }
+                // TODO: restore once the above TODO is fixed
+                //debug_assert_eq!(closer_peers.len(), closer_peers.capacity());
```
Comment (Vurich) [inline kad/src/protocol.rs]: Actually when the above is fixed then you'd want to do:

```rust
message.mut_closerPeers()
    .iter_mut()
    .map(Peer::from_peer)
    .collect::<Result<Vec<_>, _>>()?
```

Also commented code

File: kad/src/protocol.rs [hunk 3]
```diff
@@ -82,20 +82,30 @@ pub struct Peer {
...
-        let node_id = PeerId::from_bytes(peer.get_id().to_vec()).unwrap(); // TODO: don't unwrap
-        let addrs = peer.take_addrs()
-			.into_iter()
-			.map(|a| Multiaddr::from_bytes(a).unwrap())		// TODO: don't unwrap
-			.collect();
+impl Peer {
+    // Builds a `Peer` from its raw protobuf equivalent.
+    // TODO: use TryFrom once stable
+    fn from_peer(peer: &mut protobuf_structs::dht::Message_Peer) -> Result<Peer, IoError> {
+        // TODO: this is in fact a CID ; not sure if this should be handled in `from_bytes` or
```
Comment (Vurich) [inline kad/src/protocol.rs]: What's a CID?

**Review by eira-fransham — APPROVED** [review]
LGMT

File: kad/src/protocol.rs (hunk 3 above)
Comment (tomaka) [reply kad/src/protocol.rs]: https://github.com/libp2p/rust-libp2p/tree/master/cid

**Pull request merged**